### PR TITLE
MAPE

### PR DESCRIPTION
**Pull request by azev77**

### CODE REVIEW

**Comment by azev77** [comment]
@OkonSamuel I think this one has the stuff

File: src/measures/continuous.jl [hunk 1]
```diff
@@ -272,3 +272,39 @@ function (::RMSP)(ŷ::Vec{<:Real}, y::Vec{<:Real})
     end
     return sqrt(ret / count)
 end
+
+struct MAPE <: Measure end
+
+"""
+     mape(ŷ, y)
+Mean Absolute Percentage Error:
+``\\text{MAPE} =  m^{-1}∑ᵢ|{yᵢ-ŷᵢ \\over yᵢ}|``
```
Comment (OkonSamuel) [inline src/measures/continuous.jl:281]: am not too familiar with latex but shouldn't there be brackets here like ``\\text{MAPE} =  m^{-1}∑ᵢ|{(yᵢ-ŷᵢ )\\over yᵢ}|``

Comment (azev77) [reply src/measures/continuous.jl:281]: @OkonSamuel  yea

Comment (OkonSamuel) [reply src/measures/continuous.jl:281]: let me help you edit that

**Comment by tlienart** [comment]
ah crap I commented on the wrong PR, please see comments in #293 there's a few numerical issues to sort out.

**Comment by azev77** [comment]
@tlienart 
1 added `@inbounds`.
This can be added other places in the same program (I don't wanna touch other people's code)

2 regarding a tolerance, this is an issue w/ any measure based on percentage errors #95 
(currently RMSP does the same thing)
New code:
```julia
    @inbounds for i in eachindex(y)
        ayi = abs(y[i])
        if ayi > tol
        #if y[i] != zero(eltype(y))
            dev = abs((y[i] - ŷ[i]) / ayi)
            #dev = abs((y[i] - ŷ[i]) / y[i])
            ret += dev
            count += 1
        end
    end
```

3 regarding the corner case where `count = 0`, that is an inevitable [known problem](https://robjhyndman.com/papers/mase.pdf) w/ any percentage based score-measure (including RMSP).

**Comment by tlienart** [comment]
Great thanks, @ablaom will probably complain on how you pass the tolerance parameter, it should be part of the MAPE struct.

Re inevitable errors etc, I would suggest showing a warning? 

Re improving RMSP to also have a threshold + use inbounds, we should do it 👍🏼

**Comment by azev77** [comment]
None of the current measures have a tolerance parameter as part of the struct. 
How can this be done?
I don't know how to throw a warning...

**Comment by azev77** [comment]
perhaps the current code could be written more parsimoniously using mean(), median etc? 
(unless loops are faster?)
Define vector: **e := ŷ - y**
image

Define vector: **p := e/y**  {for y[i]>tol}
image

etcetera?

**Comment by tlienart** [comment]
I think @ablaom can better comment on this PR, he’s the one who thought carefully about measures.

I think the PR is great.

There’s a wider discussion of performance, warnings etc but it can be done separately. And yes afaik `mean` is two-three times faster than the loop, possibly bc it uses SIMD (?) , not sure.

Warning can be shown with `@warn` but better to do this for everything rather than have only one metric that would do it, so could be a separate issue.

**Comment by azev77** [comment]
> Ps; performance is not really a problem with metrics...
I fully agree, that's why I wasn't sure about the need for `@inbounds`, but why not...

**Comment by azev77** [comment]
@ablaom @tlienart see how much easier this could be: 
(the only things missing are weights & tol)
```julia
function err(ŷ, y)
    return ŷ - y
end

function perr(ŷ, y; tol = eps())
    e = err(ŷ, y)
    p = 100 * e[y .!= 0] ./ y[y .!= 0]
    return p
end

using Statistics
mse(ŷ, y)  = err(ŷ, y).^2 |> mean
rmse(ŷ, y) = err(ŷ, y).^2 |> mean |> sqrt
mae(ŷ, y)  = abs.(err(ŷ, y)) |> mean
mdae(ŷ, y) = abs.(err(ŷ, y)) |> median
rmdse(ŷ, y) = err(ŷ, y).^2 |> median |> sqrt

mape(ŷ, y)    = abs.(perr(ŷ, y)) |> mean
mdape(ŷ, y)   = abs.(perr(ŷ, y)) |> median
rmspe(ŷ, y)   = perr(ŷ, y).^2 |> mean |> sqrt
rmdspe(ŷ, y)  = perr(ŷ, y).^2 |> median |> sqrt

"https://en.wikipedia.org/wiki/Symmetric_mean_absolute_percentage_error"
smape(ŷ, y)   = 200. * abs.(err(ŷ, y)) ./ (ŷ + y) |> mean
smdape(ŷ, y)   = 200. * abs.(err(ŷ, y)) ./ (ŷ + y) |> median

# TEST
y    = [-1, 0, 1, 2, 3, 4]
ŷ    = [-3, 1, 2, 2, 4, 5]

e = err(ŷ, y)
p = perr(ŷ, y)

mse(ŷ, y)
rmse(ŷ, y)
mae(ŷ, y) 
mdae(ŷ, y) 
rmdse(ŷ, y)

mape(ŷ, y)
mdape(ŷ, y)
rmspe(ŷ, y)
rmdspe(ŷ, y)
smape(ŷ, y)
smdape(ŷ, y) 
```

**Comment by azev77** [comment]
I'm not sure about struct & I took out kwarg, but put in a default tol.
Can a measure have arguments w/ default values? 
```julia
struct MAPE <: Measure 
    tol::Real
end

"""
     mape(ŷ, y)
Mean Absolute Percentage Error:
``\\text{MAPE} =  m^{-1}∑ᵢ|{(yᵢ-ŷᵢ) \\over yᵢ}|`` 
where the sum is over indices such that `yᵢ≂̸0` and `m` is the number
of such indices.
For more information, run `info(mape)`.
"""
const mape = MAPE()

metadata_measure(MAPE;
    name                     = "mape",
    target_scitype           = Union{Vec{Continuous},Vec{Count}},
    prediction_type          = :deterministic,
    orientation              = :loss,
    reports_each_observation = false,
    is_feature_dependent     = false,
    supports_weights         = false,
    docstring                = "Mean Absolute Percentage Error; aliases: `mape`.")

function (::MAPE)(ŷ::Vec{<:Real}, y::Vec{<:Real}, tol = eps())
    check_dimensions(ŷ, y)
    ret = zero(eltype(y))
    count = 0
    @inbounds for i in eachindex(y)
        ayi = abs(y[i])
        if ayi > tol
        #if y[i] != zero(eltype(y))
            dev = abs((y[i] - ŷ[i]) / ayi)
            #dev = abs((y[i] - ŷ[i]) / y[i])
            ret += dev
            count += 1
        end
    end
    return ret / count
end
```

**Comment by ablaom** [comment]
No, the idea is that you choose the tolerance at time of instantiation of the measure. As in 

```julia
mape_dangerous = MAPE(tol=0)
mape_dangerous(yhat, y) 
```

The signature of the measure can only be `m(yhat, y)` or `m(yhat, y, w)`.

**Comment by azev77** [comment]
@ablaom sorry, not clear
1. can you show me w/ the mape() example 
2. include it w/ the new upgraded measures.jl (if we're gonna throw out most of the code anyway)

**Comment by ablaom** [comment]
> 1\. can you show me w/ the mape() example

See the new `mape` fork. Here's the diff: https://github.com/alan-turing-institute/MLJBase.jl/commit/9aabed73407090fc5b07b013c61748fd1c054e90

> 2\. include it w/ the new upgraded measures.jl (if we're gonna throw out most of the code anyway)

Are you volunteering to carry out the refactor at #299, then? I'm not going to get around to that for some time.

**Comment by azev77** [comment]
1. thanks for the code 
2. while I'd love to help w/ the refactor I don't know enough about the details of MLJ. 
I spilled my guts [here](https://github.com/alan-turing-institute/MLJBase.jl/issues/95#issuecomment-626411420).

**Comment by ablaom** [comment]
Closed in favour of #302 - essentially identical

**Comment by ablaom** [comment]
@azev77 Many thanks for the PR.

**Pull request closed without merging**